PR Test (#1)
Luanf/pt-BR.reactjs.org: content/docs/implementation-notes.md

  - "contributing/implementation-notes.html"
---

> Essa seção faz parte de um conjunto de notas de implementação para o [reconciliador de pilha](/docs/codebase-overview.html#stack-reconciler).

Author: ```suggestion
Esta seção é um conjunto de notas de implementação para o [reconciliador de pilha](/docs/codebase-overview.html#stack-reconciler).
```

> Ela é bastante técnica e assume um forte entendimento da API pública do React, assim como da sua divisão em núcleos, renderizadores e o reconciliador. Se você não estiver muito familiarizado com o código do React, leia a [visão geral da base de código](/docs/codebase-overview.html) primeiro.

Author: ```suggestion
Ela é bastante técnica e assume um forte entendimento da API pública do React, assim como da sua divisão em núcleos, renderizadores e o próprio reconciliador. Se você não estiver muito familiarizado com o código do React, leia a [visão geral da base de código](/docs/codebase-overview.html) primeiro.
```

> também é assumido o entendimento da [diferença entre componentes React, suas instâncias e elementos](/blog/2015/12/18/react-components-elements-and-instances.html).

Author: Também é pressuposto o entendimento da [diferença entre componentes React, suas instâncias e elementos]

> também é assumido o entendimento da [diferença entre componentes React, suas instâncias e elementos](/blog/2015/12/18/react-components-elements-and-instances.html).

Author: ```suggestion
Também é pressuposto o entendimento da [diferença entre componentes React, suas instâncias e elementos](/blog/2015/12/18/react-components-elements-and-instances.html).
```

> O reconciliador de pilha foi usado no React 15 e até anteriormente. Esta localizado em[src/renderers/shared/stack/reconciler](https://github.com/facebook/react/tree/15-stable/src/renderers/shared/stack/reconciler).

Author: ```suggestion
O reconciliador de pilha foi usado no React 15 e até anteriormente. Fica localizado em [src/renderers/shared/stack/reconciler](https://github.com/facebook/react/tree/15-stable/src/renderers/shared/stack/reconciler).
```

### Video: Construindo React do zero {#video-building-react-from-scratch}

[Paul O'Shannessy](https://twitter.com/zpao) deu uma palestra sobre [construir React do zero](https://www.youtube.com/watch?v=_MAD4Oly9yg) que muito inspirou esse documento.

> Tanto esse texto quanto a palestra são simplificações da verdadeira base de código, então se familiarizar com os dois pode resultar em um entendimento melhor.

Author: ```suggestion
Tanto esse texto quanto a palestra são simplificações da real base de código, então se familiarizar com os dois pode resultar em um entendimento melhor.
```

### Visão geral {#overview}

> O reconciliador em si não possui uma API pública. [Renderizadores](/docs/codebase-overview.html#stack-renderers) como o React DOM e React Nativa o usam para atualizar a interface do usuário de acordo com os componentes React escritos pelo usuário.

Author: ```suggestion
O reconciliador em si não possui uma API pública. [Renderizadores](/docs/codebase-overview.html#stack-renderers) como o React DOM e React Native o usam para atualizar a interface do usuário de acordo com os componentes React escritos pelo usuário.
```

> ### Montar como um Processo Recursivo {#mounting-as-a-recursive-process}

Author: ```suggestion
### Montagem como um Processo Recursivo {#mounting-as-a-recursive-process}
```

Vamos considerar a primeira vez que você monta um componente:

```js
ReactDOM.render(<App />, rootEl);
```

> React DOM ira passar `<App />` para o reconciliador. Lembre que `<App />` é um elemento React, isto é, uma descrição *do que* renderizar. Você pode pensar nele como um objeto simples:

Author: ```suggestion
O React DOM passara `<App />` para o reconciliador. Lembre-se que `<App />` é um elemento React, isto é, uma descrição *do que* renderizar. Você pode pensar nele como um objeto simples:
```

> O React DOM passara `<App />` para o reconciliador. Lembre-se que `<App />` é um elemento React, isto é, uma descrição *do que* renderizar. Você pode pensar nele como um objeto simples:

Author: ```suggestion
O React DOM passará `<App />` para o reconciliador. Lembre-se que `<App />` é um elemento React, isto é, uma descrição *do que* renderizar. Você pode pensar nele como um objeto simples:
```

```js
console.log(<App />);
// { type: App, props: {} }
```

> O reconciliador irá checar se `App` é uma classe ou uma função.

Author: ```suggestion
O reconciliador irá verificar se `App` é uma classe ou uma função.
```

> Se `App` for uma função, o reconciliador irá chamar `App(props)` para obter o elemento renderizado.

Author: ```suggestion
Se `App` for uma função, o reconciliador chama `App(props)` para obter o elemento renderizado.
```

> Se `App` for uma classe, o reconciliador irá instanciar um `App` com `new App(props)`, chamar o método de ciclo de vida `componentWillMount()`, e então irá chamar o método render()` para obter o elemento renderizado.

Author: ```suggestion
Se `App` for uma classe, o reconciliador instancia `App` com `new App(props)`, chama o método de ciclo de vida `componentWillMount()`, e então chama o método render()` para obter o elemento renderizado.
```

De qualquer forma, o reconciliador irá saber o que o elemento `App` se "renderizou a".

Esse processo é recursivo. `App` talvez se renderize a um `<Greeting />`, `Greeting` talvez se renderize a um `<Button />`, e assim em diante. O reconciliador irá “perfurar” os componentes definidos pelo usuário recursivamente enquanto ele aprende ao que cada um se renderiza.

Você pode imaginar esse processo como um pseudo-código:

```js
function isClass(type) {
  // Subclasses React.Component possuem essa flag
  return (
    Boolean(type.prototype) &&
    Boolean(type.prototype.isReactComponent)
  );
}

// Essa função recebe um elemento React (e.g. <App />)
// e retorna um DOM ou nó Nativo representando a árvore montada.
function mount(element) {
  var type = element.type;
  var props = element.props;

  // Nós vamos determinar o elemento renderizado

>   // por executar o tipo como função

Author: ```suggestion
  // executando o tipo como função
```

  // ou criando uma instância e chamando render().
  var renderedElement;
  if (isClass(type)) {
    // Componente de classe
    var publicInstance = new type(props);

>     // Seta as props

Author: ```suggestion
    // Define as props
```

    publicInstance.props = props;
    // Chama o ciclo de vida se necessário
    if (publicInstance.componentWillMount) {
      publicInstance.componentWillMount();
    }

>     // Obtêm o elemento renderizado por chamar render()

Author: ```suggestion
    // Obtêm o elemento renderizado ao chamar render()
```

    renderedElement = publicInstance.render();
  } else {
    // Componente de função
    renderedElement = type(props);
  }

  // Esse processo é recursivo pois um componente pode
  // retornar um elemento com o tipo de outro componente.
  return mount(renderedElement);

>   // Nota: essa implementação é incompleta e recursa infinitamente!

Author: ```suggestion
  // Nota: essa implementação é incompleta e recorre infinitamente!
```

>**Note:**
>
>Isso realmente *é* um pseudocódigo. Não é semelhante a implementação real. Causará um estouro de pilha porque não discutimos quando parar a recursão.

Let's recap a few key ideas in the example above:

* Os elementos do React são objetos simples que representam o tipo do componente (e.g. `App`) e as props.
* Componentes definidos pelo usuário (e.g. `App`) podem ser classes ou funções mas todos eles “se renderizam” a um elemento.
* "Montagem" é um processo recursivo que cria uma árvore DOM ou Nativa dado um elemento React de nível superior (e.g. `<App />`).

### Montando Elementos Host {#mounting-host-elements}

Esse processo seria inútil se o resultado não fosse renderizar algo na tela.

Além dos componentes definidos pelo usuário ("compostos"), elementos React podem também representar componentes para plataformas específicas ("hospedeiros"). Por exemplo, `Button` pode retornar uma `<div />` no seu método render.

Se a propriedade `type` for uma string, estamos lidando com um elemento hospedeiro:

```js
console.log(<div />);
// { type: 'div', props: {} }
```

Não há código definido pelo usuário associado com elementos do tipo hospedeiro.

> Quando o reconciliador encontra um elemento hospedeiro, ele permite que o renderizador cuide da montagem. Por exemplo, o React DOM criaria um nó DOM.

Author: ```suggestion
Quando o reconciliador encontra um elemento hospedeiro, ele permite que o renderizador cuide da montagem. Por exemplo, o React DOM criaria um nó do DOM.
```

Se o elemento hospedeiro possuir filhos, o reconciliador recursivamente os monta seguindo o mesmo algoritmo descrito acima. Não importa se os filhos são hospedeiros (como `<div><hr /></div>`) ou se são compostos (como `<div><Button /></div>`), ou os dois.

Os nós DOM produzidos pelos componentes filhos serão anexados ao nó DOM pai, e, recursivamente, a completa estrutura DOM será construída.

>**Note:**
>

> >O reconciliador em si não está ligado ao DOM. O exato resultado da montagem (por vezes chamada de "mount image" no código fonte) depende do renderizador, e pode ser um nó DOM (React DOM), uma string (React DOM Server), ou um número representando uma native view (React Native).

Author: ```suggestion
>O reconciliador em si não está ligado ao DOM. O exato resultado da montagem (por vezes chamada de "mount image" no código fonte) depende do renderizador, e pode ser um nó do DOM (React DOM), uma string (React DOM Server), ou um número representando uma native view (React Native).
```

>**Note:**
>

> >Isso realmente *é* um pseudocódigo. Não é semelhante a implementação real. Vai ser causado um stack overflow pois não discutimos quando parar a recursão.

Author: ```suggestion
>Isso realmente *é* um pseudocódigo. Não é semelhante a implementação real. Causará um estouro de pilha porque não discutimos quando parar a recursão.
```

> 

Author: ```suggestion
```

Let's recap a few key ideas in the example above:

> * Elementos do React são objetos planos que representam o tipo do componente (e.g. `App`) e o props.

Author: ```suggestion
* Os elementos do React são objetos simples que representam o tipo do componente (e.g. `App`) e as props.
```

* Componentes definidos pelo usuário (e.g. `App`) podem ser classes ou funções mas todos eles “se renderizam” a um elemento.

> * "Montar" é um processo recursivo que cria uma árvore DOM ou Nativa dado um elemento React de alto nível (e.g. `<App />`).

Author: ```suggestion
* "Montagem" é um processo recursivo que cria uma árvore DOM ou Nativa dado um elemento React de nível superior (e.g. `<App />`).
```

### Montando Elementos Host {#mounting-host-elements}

Esse processo seria inútil se o resultado não fosse renderizar algo na tela.

> Em adição aos componentes definidos pelo usuário ("compostos"), elementos React podem também representar componentes para plataformas específicas ("host"). Por exemplo, `Button` pode retornar um `<div />` do seu método render.

Author: ```suggestion
Além dos componentes definidos pelo usuário ("compostos"), elementos React podem também representar componentes para plataformas específicas ("hospedeiros"). Por exemplo, `Button` pode retornar uma `<div />` no seu método render.
```

> Se a propriedade `type` for uma string, estamos lidando com um elemento host:

Author: ```suggestion
Se a propriedade `type` for uma string, estamos lidando com um elemento hospedeiro:
```

```js
console.log(<div />);
// { type: 'div', props: {} }
```

> Não há código definido pelo usuário associado com elementos host.

Author: ```suggestion
Não há código definido pelo usuário associado com elementos do tipo hospedeiro.
```

> Quando o reconciliador encontra um elemento host, ele permite que o renderizador cuide de o montar. Por exemplo, o React DOM criaria um nó DOM.

Author: ```suggestion
Quando o reconciliador encontra um elemento hospedeiro, ele permite que o renderizador cuide da montagem. Por exemplo, o React DOM criaria um nó DOM.
```

> When the reconciliador encounters a host element, it lets the renderer take care of mounting it. For example, React DOM would create a DOM node.

Author: ```suggestion
```

> Se o elemento host possuir filhos, o reconciliador recursivamente os monta seguindo o mesmo algoritmo descrito acima. Não importa se os filhos são hosts (como `<div><hr /></div>`) ou se são composite (como `<div><Button /></div>`), ou os dois.

Author: ```suggestion
Se o elemento hospedeiro possuir filhos, o reconciliador recursivamente os monta seguindo o mesmo algoritmo descrito acima. Não importa se os filhos são hospedeiros (como `<div><hr /></div>`) ou se são compostos (como `<div><Button /></div>`), ou os dois.
```

> Os nós DOM produzidos pelos componentes filhos serão anexados ao nó DOM pai, e recursivamente, a completa estrutura DOM será construída.

Author: ```suggestion
Os nós DOM produzidos pelos componentes filhos serão anexados ao nó DOM pai, e, recursivamente, a completa estrutura DOM será construída.
```

>**Note:**
>

> >O reconciliador em si não está ligado ao DOM. O exato resultado da montagem (por vezes chamada de "mount image" no código fonte) depende do renderizador, e pode ser um nó DOM (React DOM), uma string (React DOM Server), ou um número representando uma native view (React Nativa).

Author: ```suggestion
>O reconciliador em si não está ligado ao DOM. O exato resultado da montagem (por vezes chamada de "mount image" no código fonte) depende do renderizador, e pode ser um nó DOM (React DOM), uma string (React DOM Server), ou um número representando uma native view (React Native).
```

> Se fossemos extender o código para lidar com elementos host, ficaria assim:

Author: ```suggestion
Se fossemos extender o código para lidar com elementos hospedeiros, ficaria assim:
```

```js
function isClass(type) {
  // Subclasses React.Component possuem essa flag
  return (
    Boolean(type.prototype) &&
    Boolean(type.prototype.isReactComponent)
  );
}

> // Essa função apenas lida com elementos do tipo composite.

Author: ```suggestion
// Essa função apenas lida com elementos do tipo composto.
```

// Por exemplo, ela lida com <App /> e <Button />, mas não com uma <div />.
function mountComposite(element) {
  var type = element.type;
  var props = element.props;

  var renderedElement;
  if (isClass(type)) {
    // Componente de classe
    var publicInstance = new type(props);

>     // Seta as props

Author: ```suggestion
    // Define as props
```

    publicInstance.props = props;
    // Chama o ciclo de vida se necessário
    if (publicInstance.componentWillMount) {
      publicInstance.componentWillMount();
    }
    renderedElement = publicInstance.render();
  } else if (typeof type === 'function') {
    // Componente de função
    renderedElement = type(props);
  }

  // Isso é recursivo mas eventualmente chegaremos no fim da recursão quando

>   // o elemento é host (e.g. <div />) ao invés de composite (e.g. <App />):

Author: ```suggestion
  // o elemento é hospedeiro (e.g. <div />) ao invés de composto (e.g. <App />):
```

  return mount(renderedElement);
}

> // Essa função apenas lida com elementos do tipo host.

Author: ```suggestion
// Essa função apenas lida com elementos do tipo hospedeiro.
```

  }
  children = children.filter(Boolean);

  // Esse bloco de código não deveria estar no reconciliador.
  // Renderizadores diferentes podem inicializar nós diferentemente.
  // Por exemplo, React Native iria criar views de iOS ou Android.
  var node = document.createElement(type);
  Object.keys(props).forEach(propName => {
    if (propName !== 'children') {
      node.setAttribute(propName, props[propName]);
    }
  });

  // Monta os filhos
  children.forEach(childElement => {
    // Filhos podem ser hospedeiros (e.g. <div />) ou compostos (e.g <Button />).
    // também os montaremos recursivamente:
    var childNode = mount(childElement);

    // Essa linha de código também é específica do renderizador.
    // Ela seria diferente dependendo do renderizador:
    node.appendChild(childNode);
  });

>   // Retorna o nó DOM como resultado da montagem.

Author: ```suggestion
  // Retorna o nó do DOM como resultado da montagem.
```

  }
  children = children.filter(Boolean);

  // Esse bloco de código não deveria estar no reconciliador.

>   // Renderizadores diferentes podem inicializar nos diferentemente.

Author: ```suggestion
  // Renderizadores diferentes podem inicializar nós diferentemente.
```

  // Por exemplo, React Native iria criar views de iOS ou Android.
  var node = document.createElement(type);
  Object.keys(props).forEach(propName => {
    if (propName !== 'children') {
      node.setAttribute(propName, props[propName]);
    }
  });

  // Monta os filhos
  children.forEach(childElement => {

>     // Filhos podem ser host (e.g. <div />) ou composite (e.g <Button />).

Author: ```suggestion
    // Filhos podem ser hospedeiros (e.g. <div />) ou compostos (e.g <Button />).
```

rootEl.appendChild(node);
```

> Isso está funcionando mas ainda está longe de como o reconciliador é realmente implementado. O ingrediente faltante é o suporte para atualizações.

Author: ```suggestion
Isso está funcionando mas ainda está longe de como o reconciliador é realmente implementado. O ingrediente que falta é o suporte para atualizações.
```

### Introduzindo instâncias Internas {#introducing-internal-instances}

> 

Author: ```suggestion
```

> A feature chave do React é que você pode rerenderizar tudo, e ele não irá recriar o DOM ou resetar o estado.

Author: ```suggestion
A feature chave do React é que você pode re-renderizar tudo, e ele não irá recriar o DOM ou resetar o estado.
```

```js
ReactDOM.render(<App />, rootEl);
// Deve reutilizar o DOM existente:
ReactDOM.render(<App />, rootEl);
```

> Contudo, nossa implementação acima apenas sabe como montar a árvore inicial. Ela não executa atualizações na árvore pois não armazena todas as informações necessárias, como todas as `publicInstance`s, ou que nos DOM correspondem a qual componente.

Author: ```suggestion
Contudo, nossa implementação acima apenas sabe como montar a árvore inicial. Ela não executa atualizações na árvore pois não armazena todas as informações necessárias, como todas as `publicInstance`s, ou que nós DOM correspondem a qual componente.
```

> O código do stack reconciliador resolve isso fazendo a função `mount()` um método e a colocando em uma classe. Existem desvantagens para essa abordagem, e nos iremos na direção oposta na [atual reescrita do reconciliador.(/docs/codebase-overview.html#fiber-reconciler). No entanto, é assim que funciona atualmente.

Author: ```suggestion
O código do reconciliador de pilha resolve isso tornando a função `mount()` em um método de classe. Existem desvantagens para essa abordagem, e nós iremos na direção oposta na [atual reescrita do reconciliador](/docs/codebase-overview.html#fiber-reconciler). No entanto, é assim que funciona atualmente.
```

Ao invés de funções `mountHost` e `mountComposite` separadas, nos criaremos duas classes: `DOMComponent` e `CompositeComponent`.

> Ambas classes possuem um construtor aceitando o `element`, assim como um método `mount()` retornando o código montado. Nós iremos trocar a função de alto nível `mount()` com uma factory que instancia a classe correta.

Author: ```suggestion
Ambas classes possuem um construtor aceitando o `element`, assim como um método `mount()` retornando o nó montado. Nós iremos trocar a função de nível superior `mount()` com uma _factory_ que instancia a classe correta.
```

rootEl.appendChild(node);
```

> Isso está funcionando mas ainda é longe de como o reconciliador está realmente implementado. O ingrediente faltante é o suporte para atualizações.

Author: ```suggestion
Isso está funcionando mas ainda está longe de como o reconciliador é realmente implementado. O ingrediente faltante é o suporte para atualizações.
```

  }

  getPublicInstance() {

>     // Para componentes composite, exponha a instância da classe.

Author: ```suggestion
    // Para componentes compostos, exponha a instância da classe.
```

    var publicInstance;
    var renderedElement;
    if (isClass(type)) {
      // Componente de classe
      publicInstance = new type(props);

>       // Seta as props

Author: ```suggestion
      // Define as props
```

      publicInstance.props = props;
      // Chama o ciclo de vida se necessário
      if (publicInstance.componentWillMount) {
        publicInstance.componentWillMount();
      }
      renderedElement = publicInstance.render();
    } else if (typeof type === 'function') {
      // Componente de função
      publicInstance = null;
      renderedElement = type(props);
    }

    // Salva a instância pública
    this.publicInstance = publicInstance;

    // Instancia a instância interna filha de acordo com o elemento.
    // Seria um DOMComponent para <div /> ou <p />,
    // e um CompositeComponent para <App /> ou <Button />:
    var renderedComponent = instantiateComponent(renderedElement);
    this.renderedComponent = renderedComponent;

    // Monta o output renderizado
    return renderedComponent.mount();
  }
}
```

> Isso não é muito diferente da nossa implementação anterior de `mountComposite`, mas agora podemos salvar alguma informação, como `this.currentElement`, `this.renderedComponent`, e `this.publicInstance` , para usar durante atualizações.

Author: ```suggestion
Isso não é muito diferente da nossa implementação anterior de `mountComposite`, mas agora podemos salvar algumas informações, como `this.currentElement`, `this.renderedComponent`, e `this.publicInstance` , para usar durante atualizações.
```

Note que uma instância de `CompositeComponent` não é a mesma coisa que uma instância de um `element.type` fornecido pelo usuário. `CompositeComponent` é um detalhe de implementação do nosso reconciliador, e nunca é exposto para o usuário. A classe definida pelo usuário é quem lê de `element.type`, e `CompositeComponent` cria uma instância dela.

> Para evitar confusão, nós vamos chamar instâncias de `CompositeComponent` e `DOMComponent` "instâncias internas". Elas existem para que possamos às associar com alguns dados de longa vida. Apenas o renderizador e o reconciliador sabem que elas existem.

Author: ```suggestion
Para evitar confusão, nós vamos chamar instâncias de `CompositeComponent` e `DOMComponent` "instâncias internas". Elas existem para que possamos às associar a alguns dados de longa vida. Apenas o renderizador e o reconciliador sabem que elas existem.
```

      children = [children];
    }

    // Cria e salva o nó
    var node = document.createElement(type);
    this.node = node;

    // Define os atributos
    Object.keys(props).forEach(propName => {
      if (propName !== 'children') {
        node.setAttribute(propName, props[propName]);
      }
    });

    // Cria e salva os filhos contidos.
    // Cada um deles pode ser um DOMComponent ou um CompositeComponent
    // dependendo se o tipo do elemento é uma string ou uma função.
    var renderedChildren = children.map(instantiateComponent);
    this.renderedChildren = renderedChildren;

    // Coleta nos DOM retornados na montagem
    var childNodes = renderedChildren.map(child => child.mount());
    childNodes.forEach(childNode => node.appendChild(childNode));

>     // Retorna o nó DOM como resultado da montagem

Author: ```suggestion
    // Retorna o nó do DOM como resultado da montagem
```

      children = [children];
    }

    // Cria e salva o nó
    var node = document.createElement(type);
    this.node = node;

>     // Seta os atributos

Author: ```suggestion
    // Define os atributos
```

    Object.keys(props).forEach(propName => {
      if (propName !== 'children') {
        node.setAttribute(propName, props[propName]);
      }
    });

    // Cria e salva os filhos contidos.
    // Cada um deles pode ser um DOMComponent ou um CompositeComponent
    // dependendo se o tipo do elemento é uma string ou uma função.
    var renderedChildren = children.map(instantiateComponent);
    this.renderedChildren = renderedChildren;

    // Coleta nos DOM retornados na montagem
    var childNodes = renderedChildren.map(child => child.mount());
    childNodes.forEach(childNode => node.appendChild(childNode));

>     // Retorna o nó DOM com resultado da montagem

Author: ```suggestion
    // Retorna o nó DOM como resultado da montagem
```

    return node;
  }
}
```

> %%% A diferença principal depois de refatorar de `mountHost()` é que agora nos podemos deixar `this.node` e `this.renderedChildren` associados com a instância interna do componente DOM. Nós também os usaremos para aplicar atualizações não destrutivas no futuro.

Author: ```suggestion
A diferença principal depois de refatorar `mountHost()` é que agora nós podemos deixar `this.node` e `this.renderedChildren` associados com a instância interna do componente DOM. Nós também os usaremos para aplicar atualizações não destrutivas no futuro.
```

> Como resultado, cada instância interna, composite ou host, agora aponta para sua instância interna filha. Para auxiliar na visualização disso, se o componente de função `<App>` renderiza um componente de classe, e a classe `Button` renderiza a `<div>`, a árvore da instância interna ficaria assim:

Author: ```suggestion
Como resultado, cada instância interna, composta ou hospedeira, agora aponta para sua instância interna filha. Para auxiliar na visualização disso, se o componente de função `<App>` renderiza um componente de classe, e a classe `Button` renderiza a `<div>`, a árvore da instância interna ficaria assim:
```

}
```

No DOM você apenas veria a `<div>`. No entanto, a árvore da instância interna possui ambas instâncias internas: composta e hospedeira.

A instância interna composta precisa armazenar:

* O elemento atual.
* A instância pública se o tipo do elemento for uma classe.
* A única instância interna renderizada. Pode ser tanto um `DOMComponent` ou um `CompositeComponent`.

A instância interna hospedeira precisa armazenar:

* O elemento atual.

> * O nó DOM.

Author: ```suggestion
* O nó do DOM.
```

}
```

> No DOM você apenas veria a `<div>`. No entanto, a árvore da instância interna possui ambas instâncias internas: composite e host.

Author: ```suggestion
No DOM você apenas veria a `<div>`. No entanto, a árvore da instância interna possui ambas instâncias internas: composta e hospedeira.
```

> A instância interna composite precisa armazenar:

Author: ```suggestion
A instância interna composta precisa armazenar:
```

* O elemento atual.
* A instância pública se o tipo do elemento for uma classe.
* A única instância interna renderizada. Pode ser tanto um `DOMComponent` ou um `CompositeComponent`.

> O instância interna host precisa armazenar:

Author: ```suggestion
A instância interna hospedeira precisa armazenar:
```

* O elemento atual.
* O nó DOM.
* Todas as instâncias internas filhas. Cada uma delas pode ser tanto um `DOMComponent` ou um `CompositeComponent`.

> Se você está tendo dificuldades para imaginar como uma árvore de instâncias internas  é estruturada em aplicações mais complexas, [React DevTools](https://github.com/facebook/react-devtools) pode te dar uma boa aproximação, pois instâncias host são marcadas com cinza, e instâncias composite com roxo:

Author: ```suggestion
Se você está tendo dificuldades para imaginar como uma árvore de instâncias internas é estruturada em aplicações mais complexas, [React DevTools](https://github.com/facebook/react-devtools) pode te dar uma boa aproximação, pois instâncias hospedeiras são marcadas com cinza, e instâncias compostas com roxo:
```

```js
class CompositeComponent {

  // ...

  unmount() {
    // Chama o método de ciclo de vida se necessário
    var publicInstance = this.publicInstance;
    if (publicInstance) {
      if (publicInstance.componentWillUnmount) {
        publicInstance.componentWillUnmount();
      }
    }

    // Desmonta o componente renderizado
    var renderedComponent = this.renderedComponent;
    renderedComponent.unmount();
  }
}
```

Para o `DOMComponent`, desmontar pede para que todo filho se desmonte:

```js
class DOMComponent {

  // ...

  unmount() {
    // Desmonta todos os filhos
    var renderedChildren = this.renderedChildren;
    renderedChildren.forEach(child => child.unmount());
  }
}
```

Na prática, desmontar componentes DOM também remove os event listeners e limpa alguns caches, mas vamos pular essas detalhes.

Podemos agora adicionar uma nova função de alto nível chamada `unmountTree(containerNode)` que é semelhante a `ReactDOM.unmountComponentAtNode()`.

```js
function unmountTree(containerNode) {

>   // Le a instância interna de um nó DOM:

Author: ```suggestion
  // Le a instância interna de um nó do DOM:
```

  // (Isso ainda não funciona, nós vamos precisar mudar mountTree() para guarda-la)
  var node = containerNode.firstChild;
  var rootComponent = node._internalInstance;

  // Desmonta a árvore e limpa o container
  rootComponent.unmount();
  containerNode.innerHTML = '';
}
```

> Para que isso funcione, nós precisamos ler uma instância interna raiz de um nó DOM. Nós vamos modificar `mountTree()` para adicionar a propriedade `_internalInstance` ao nó DOM raiz. Nós também ensinaremos a `mountTree()` como destruir qualquer árvore existente para que ela possa ser chamada múltiplas vezes:

Author: ```suggestion
Para que isso funcione, nós precisamos ler uma instância interna raiz de um nó do DOM. Nós vamos modificar `mountTree()` para adicionar a propriedade `_internalInstance` ao nó do DOM raiz. Nós também ensinaremos a `mountTree()` como destruir qualquer árvore existente para que ela possa ser chamada múltiplas vezes:
```

mountTree(<App />, rootEl);
```

### Desmontando {#unmounting}

> Agora que temos instâncias internas que possuem seus filhos e nos DOM, podemos implementar o desmontar. Para um componente composite, desmontar chama um método do ciclo de vida e recursos.

Author: ```suggestion
Agora que temos instâncias internas que possuem seus filhos e nós DOM, podemos implementar o desmontar. Para um componente composto, desmontar executa um método do ciclo de vida e recorre.
```

```js
class CompositeComponent {

  // ...

  unmount() {
    // Chama o método de ciclo de vida se necessário
    var publicInstance = this.publicInstance;
    if (publicInstance) {
      if (publicInstance.componentWillUnmount) {
        publicInstance.componentWillUnmount();
      }
    }

>     // Desmonta o singular componente renderizado

Author: ```suggestion
    // Desmonta o componente renderizado
```

    var renderedComponent = this.renderedComponent;
    renderedComponent.unmount();
  }
}
```

> Para `DOMComponent`, desmontar pede para todo filho desmontar:

Author: ```suggestion
Para o `DOMComponent`, desmontar pede para que todo filho se desmonte:
```

```js
class DOMComponent {

  // ...

  unmount() {
    // Desmonta todos os filhos
    var renderedChildren = this.renderedChildren;
    renderedChildren.forEach(child => child.unmount());
  }
}
```

> Na prática, desmontar componentes DOM também remove os event listeners e limpa alguns caches, mas vamos pular essas detalhes.

Author: ```suggestion
Na prática, desmontar componentes DOM também remove os event listeners e limpa alguns caches, mas vamos pular esses detalhes.
```

> Podemos agora adicionar uma nova função de alto nível chamada `unmountTree(containerNode)` que é similar a `ReactDOM.unmountComponentAtNode()`.

Author: ```suggestion
Podemos agora adicionar uma nova função de alto nível chamada `unmountTree(containerNode)` que é semelhante a `ReactDOM.unmountComponentAtNode()`.
```

```js
function unmountTree(containerNode) {

>   // Le a instância interna de um no DOM:

Author: ```suggestion
  // Le a instância interna de um nó DOM:
```

>   // (Isso ainda não funciona, nos vamos precisar mudar mountTree() para guarda-la)

Author: ```suggestion
  // (Isso ainda não funciona, nós vamos precisar mudar mountTree() para guarda-la)
```

  var node = containerNode.firstChild;
  var rootComponent = node._internalInstance;

  // Desmonta a árvore e limpa o container
  rootComponent.unmount();
  containerNode.innerHTML = '';
}
```

> Para que isso funcione, nós precisamos ler uma instância interna raiz de um no DOM. Nós vamos modificar `mountTree()` para adicionar a propriedade `_internalInstance` ao nó DOM raiz. Nós também ensinaremos a `mountTree()` como destruir qualquer árvore existente para que ela possa ser chamada múltiplas vezes:

Author: ```suggestion
Para que isso funcione, nós precisamos ler uma instância interna raiz de um nó DOM. Nós vamos modificar `mountTree()` para adicionar a propriedade `_internalInstance` ao nó DOM raiz. Nós também ensinaremos a `mountTree()` como destruir qualquer árvore existente para que ela possa ser chamada múltiplas vezes:
```

```js
function mountTree(element, containerNode) {
  // Destrói qualquer árvore existente
  if (containerNode.firstChild) {
    unmountTree(containerNode);
  }

  // Cria a instância interna de nível superior
  var rootComponent = instantiateComponent(element);

  // Monta o componente de nivel superior no container
  var node = rootComponent.mount();
  containerNode.appendChild(node);

  // Salva a referência a instância interna
  node._internalInstance = rootComponent;

  // Retorna a instância pública que é provida
  var publicInstance = rootComponent.getPublicInstance();
  return publicInstance;
}
```

Agora, executar `unmountTree()` ou executar `mountTree()` repetidamente, remove a árvore antiga e executa o método de ciclo de vida `componentWillUnmount()` nos componentes.

### Atualizando {#updating}

> Na seção anterior, nos implementamos o desmontar. Contudo, o React não seria muito útil se cada mudança de prop desmontasse e montasse a árvore toda. O objetivo do reconciliador é reusar instâncias existentes quando possível para preservar o DOM e o estado:

Author: ```suggestion
Na seção anterior, nós implementamos o desmontar. Contudo, o React não seria muito útil se cada mudança de prop desmontasse e montasse a árvore toda. O objetivo do reconciliador é reutilizar instâncias existentes quando possível para preservar o DOM e o estado:
```

```js
var rootEl = document.getElementById('root');

mountTree(<App />, rootEl);
// Deve reutilizar o DOM existente:
mountTree(<App />, rootEl);
```

Nós iremos estender esse contrato da instância interna com mais um método. Alem do `mount()` e `unmount()`, ambos `DOMComponent` e `CompositeComponent` vão implementar um novo método chamado `receive(nextElement)`:

> We will extend our internal instance contract with one more method. In addition to `mount()` and `unmount()`, both `DOMComponent` and `CompositeComponent` will implement a new method called `receive(nextElement)`:

Author: ```suggestion
```

}
```

Sua responsabilidade é fazer o que for necessário para atualizar o componente (e qualquer um de seus filhos) com a descrição dada pelo `nextElement`.

Essa é a parte geralmente descrita como "diff do virtual DOM" embora o que realmente acontece é que andamos pela árvore interna recursivamente e permitimos que cada instância receba uma atualização.

> ### Atualizando elemento composite {#updating-composite-components}

Author: ```suggestion
### Atualizando Elementos Compostos {#updating-composite-components}
```

> Quando um componente composite recebe um novo elemento, nos rodamos o método de ciclo de vida `componentWillUpdate()`.

Author: ```suggestion
Quando um componente composto recebe um novo elemento, nós executamos o método de ciclo de vida `componentWillUpdate()`.
```

> Então rerenderizamos o componente com a nova props, e pegamos o proximo elemento renderizado:

Author: ```suggestion
Então re-renderizamos o componente com as novas props, e capturamos o próximo elemento renderizado:
```

    var prevRenderedComponent = this.renderedComponent;
    var prevRenderedElement = prevRenderedComponent.currentElement;

    // Atualiza *o proprio* elemento
    this.currentElement = nextElement;
    var type = nextElement.type;
    var nextProps = nextElement.props;

    // Descobre qual é o próximo output de render()
    var nextRenderedElement;
    if (isClass(type)) {
      // Componente de classe
      // Chama o ciclo de vida se necessário
      if (publicInstance.componentWillUpdate) {
        publicInstance.componentWillUpdate(nextProps);
      }
      // Atualiza as props
      publicInstance.props = nextProps;

>       // Rerenderiza

Author: ```suggestion
      // Re-renderiza
```

      nextRenderedElement = publicInstance.render();
    } else if (typeof type === 'function') {
      // Componente de função
      nextRenderedElement = type(nextProps);
    }

    // ...
```

> Após isso, nós podemos olhar para o `type` do elemento renderizado. Se o `type` não mudou desde a última renderização, o componente abaixo tanto pode ser atualizado %%.

Author: ```suggestion
Após isso, nós podemos olhar para o `type` do elemento renderizado. Se o `type` não mudou desde a última renderização, o componente abaixo também pode ser atualizado.
```

%% Por exemplo, se retornado `<Button color="red" />` na primeira vez, e `<Button color="blue" />` na segunda vez, nós podemos apenas dizer a instância interna correspondente à `receive()` o próximo elemento:

For example, if it returned `<Button color="red" />` the first time, and `<Button color="blue" />` the second time, we can just tell the corresponding internal instance to `receive()` the next element:

```js
    // ...

    // Se o tipo do elemento renderizado não mudou

>     // reutiliza a instância existente do componente e retorne.

Author: ```suggestion
    // reutilize a instância existente do componente e retorne.
```

    // ...
```

Contudo, se o próximo elemento renderizado possuir um `type` diferente do anterior, não podemos atualizar a instância interna. Um `<button>` não pode "se tornar" um `<input>`.

> Nesse caso, temos que desmontar a instância interna existente e montar a nova correspondente ao tipo do elemento renderizado. Por exemplo, é isso que acontece quando um componente que previamente renderizada um `<button />` renderiza um `<input />`:

Author: ```suggestion
Nesse caso, temos que desmontar a instância interna existente e montar a nova correspondente ao tipo do elemento renderizado. Por exemplo, é isso que acontece quando um componente que previamente renderizava um `<button />` renderiza um `<input />`:
```

```js
    // ...

>     // Se chegamos nesse ponto, nos precisamos desmontar o componente

Author: ```suggestion
    // Se chegamos nesse ponto, nós precisamos desmontar o componente
```

>     // montado anteriormente, montar o novo, e trocar seus nos.

Author: ```suggestion
    // montado anteriormente, montar o novo, e trocar seus nós.
```

>     // Encontra o nó antigo pois sera necessário troca-lo

Author: ```suggestion
    // Encontra o nó antigo pois será necessário troca-lo
```

    var prevNode = prevRenderedComponent.getHostNode();

    // Desmonta o filho antigo e monta o novo
    prevRenderedComponent.unmount();
    var nextRenderedComponent = instantiateComponent(nextRenderedElement);
    var nextNode = nextRenderedComponent.mount();

    // Substitui a referencia ao filho
    this.renderedComponent = nextRenderedComponent;

    // Substitui o nó antigo com o novo

>     // Nota: isso é código específico do renderizado e

Author: ```suggestion
    // Nota: isso é código específico do renderizador e
```

    // idealmente deveria viver fora do CompositeComponent
    prevNode.parentNode.replaceChild(nextNode, prevNode);
  }
}
```

> Para resumir isso tudo, quando um componente composite recebe um novo elemento, ele pode ou não delegar a atualização a sua instância interna renderizada, ou a desmontar e montar uma nova em seu lugar.

Author: ```suggestion
Para resumir isso tudo, quando um componente composto recebe um novo elemento, ele pode ou não delegar a atualização a sua instância interna renderizada, ou a desmontar e montar uma nova em seu lugar.
```

> Existe outra condição na qual um componente vai remontar ao invés de receber um elemento, e isso é quando a chave do elemento mudou. Nós não discutimos sobre como lidar com chaves nesse documento pois adiciona mais complexidade a um tutorial já complexo.

Author: ```suggestion
Existe outra condição na qual um componente vai remontar ao invés de receber um elemento, e isso é quando a chave do elemento mudou. Nós não discutimos sobre como lidar com chaves nesse documento pois adicionaria mais complexidade a um tutorial já complexo.
```

> Note que nós precisamos adicionar um método chamado `getHostNode()` para a o contrato de uma instância interna para que ela possa localizar o nó platform-specific e o trocar durante a atualização. Sua implementação é bem direta para ambas as classes:

Author: ```suggestion
Note que nós precisamos adicionar um método chamado `getHostNode()` ao contrato de uma instância interna para que ela possa localizar o nó específico de plataforma e o trocar durante a atualização. Sua implementação é bem direta para ambas as classes:
```

```js
class CompositeComponent {
  // ...

  getHostNode() {
    // Peça ao componente renderizado para fornecê-lo.

>     // Isso irá acessar recursivamente quaisquer elementos composite.

Author: ```suggestion
    // Isso irá acessar recursivamente quaisquer elementos compostos.
```

}
```

> ### Atualizando Componentes Host {#updating-host-components}

Author: ```suggestion
### Atualizando Componentes Hospedeiros {#updating-host-components}
```

> Implementações de componentes host, como a de `DOMComponent`, atualizam de maneira diferente. Quando recebem um elemento, é preciso atualizar a platform-specific view subjacente. No caso de React DOM, isso significa atualizar os atributos DOM:

Author: ```suggestion
Implementações de componentes hospedeiros, como a de `DOMComponent`, atualizam de maneira diferente. Quando recebem um elemento, é preciso atualizar a view específica de plataforma subjacente. No caso de React DOM, isso significa atualizar os atributos DOM:
```

    var nextProps = nextElement.props;    
    this.currentElement = nextElement;

    // Remove atributos antigos
    Object.keys(prevProps).forEach(propName => {
      if (propName !== 'children' && !nextProps.hasOwnProperty(propName)) {
        node.removeAttribute(propName);
      }
    });

>     // Seta os próximos atributos.

Author: ```suggestion
    // Define os próximos atributos.
```

    // ...
```

> Então, o componente host precisa atualizar seus filhos. Diferentemente de componentes composite, eles podem conter mais de um filho.

Author: ```suggestion
Então, o componente hospedeiro precisa atualizar seus filhos. Diferentemente de componentes compostos, eles podem conter mais de um filho.
```

Nesse seguinte exemplo simplificado, nos usamos um array de instâncias internas e iteramos sobre ele, atualizando ou trocando as instâncias internas, dependendo se o `type` recebido é igual ao `type` anterior. O verdadeiro reconciliador também toma a chave do elemento em conta e rastreia movimentos, além de inserções e deletes, mas omitiremos essa lógica.

> Nós coletamos operações DOM nós filhos em uma lista para que possamos executá-las em lotes:

Author: ```suggestion
Nós coletamos operações DOM feitas em nós filhos em uma lista para que possamos executá-las em lotes:
```

    if (!Array.isArray(nextChildren)) {
      nextChildren = [nextChildren];
    }
    // Esses são vetores de instâncias internas:
    var prevRenderedChildren = this.renderedChildren;
    var nextRenderedChildren = [];

    // À medida que iteramos os filhos, adicionaremos operações ao vetor.
    var operationQueue = [];
 
    // Nota: a secção abaixo é extremamente simplificada!
    // Ela não lida com re-ordenações, filhos com furos, ou chaves.
    // Ela só existe para ilustrar o fluxo geral, não os detalhes.

    for (var i = 0; i < nextChildren.length; i++) {
      // Tenta obter uma instância interna existente para esse filho
      var prevChild = prevRenderedChildren[i];

      // Se não houver instâncias internas nesse índice,

>       // um filho foi anexo ao fim. Cria uma nova instância

Author: ```suggestion
      // um filho foi anexado até o fim. Cria uma nova instância
```

      // interna, a monta, e usa seu nó.
      if (!prevChild) {
        var nextChild = instantiateComponent(nextChildren[i]);
        var node = nextChild.mount();

        // Grava que precisamos anexar um nó
        operationQueue.push({type: 'ADD', node});
        nextRenderedChildren.push(nextChild);
        continue;
      }

>       // Nós podemos apenas atualizar a instância se o tipo do elemento for compatível.

Author: ```suggestion
      // Nós podemos atualizar a instância apenas se o tipo do elemento for compatível.
```

      // Por exemplo, <Button size="small" /> pode ser atualizado a
      // <Button size="large" /> mas não a um <App />.
      var canUpdate = prevChildren[i].type === nextChildren[i].type;

      // Se não podemos atualizar uma instância existente, temos que desmonta-la

>       // e montar uma nova ao invés dela.

Author: ```suggestion
      // e montar uma nova no seu lugar.
```

      if (!canUpdate) {
        var prevNode = prevChild.getHostNode();
        prevChild.unmount();

        var nextChild = instantiateComponent(nextChildren[i]);
        var nextNode = nextChild.mount();

>         // Lembre que precisamos trocar os nos.

Author: ```suggestion
        // Lembre que precisamos trocar os nós.
```

}
```

> E é isso para atualizar componentes host.

Author: ```suggestion
E é isso para atualizar componentes hospedeiros.
```

> ### Atualizações de Alto nível {#top-level-updates}

Author: ```suggestion
### Atualizações de Nível Superior {#top-level-updates}
```

> Agora que ambos `CompositeComponent` e `DOMComponent` implementam o método `receive(nextElement)`, podemos mudar a função `mountTree()` de alto nível para que seja usada quando o tipo do elemento for o mesmo da última vez:

Author: ```suggestion
Agora que ambos `CompositeComponent` e `DOMComponent` implementam o método `receive(nextElement)`, podemos mudar a função `mountTree()` de nível superior para que seja usada quando o tipo do elemento for o mesmo da última vez:
```

```js
function mountTree(element, containerNode) {

>   // Cheque para uma árvore existente

Author: ```suggestion
  // Verifica se há uma árvore existente
```

  if (containerNode.firstChild) {
    var prevNode = containerNode.firstChild;
    var prevRootComponent = prevNode._internalInstance;
    var prevElement = prevRootComponent.currentElement;

    // Se pudermos, reutilizamos o componente raíz existente
    if (prevElement.type === element.type) {
      prevRootComponent.receive(element);
      return;
    }

>     //

Author: ```suggestion
```

>     // De outra forma, desmonte a árvore existente.

Author: ```suggestion
    // Do contrário, desmonte a árvore existente.
```

}
```

Agora chamar `mountTree()` duas vezes com o mesmo tipo não é destrutivo:

```js
var rootEl = document.getElementById('root');

mountTree(<App />, rootEl);
// Reutiliza o DOM existente:
mountTree(<App />, rootEl);
```

Esse é o básico de como o React funciona internamente.

### O Que Omitimos {#what-we-left-out}

> Esse documento é simples comparado com o código real. Existem alguns aspectos importantes que não endereçamos:

Author: ```suggestion
Esse documento é simples comparado com o código real. Existem alguns aspectos importantes que não abordamos:
```

> * Componentes podem renderizar `null`, e o reconciliador pode lidar "espaços vazios" em vetores e output renderizado.

Author: ```suggestion
* Componentes podem renderizar `null`, e o reconciliador pode lidar com "espaços vazios" em vetores em output renderizado.
```

> * O reconciliador também lê a chave de seus elementos, e a usa para estabelecer qual instância interna corresponde a qual elemento em um array. Muita da complexidade da implementação real do React esta relacionado a isto.

Author: ```suggestion
* O reconciliador também lê a chave de seus elementos, e a usa para estabelecer qual instância interna corresponde a qual elemento em um array. Muita da complexidade da implementação real do React está relacionado a isto.
```

> * Além da classe de instância interna composite e host, existem também classes para componentes texto e componentes vazios. Eles representam nos textuais e os "espaços vazios" você consegue por renderizar `null`.

Author: ```suggestion
* Além da classe de instância interna composta e hospedeira, existem também classes para componentes texto e componentes vazios. Eles representam nós textuais e os "espaços vazios" você consegue por renderizar `null`.
```

> * Renderizadores usam [injeção](/docs/codebase-overview.html#dynamic-injection) para passar a classe interna do host ao reconciliador. Por exemplo, o React DOM pede para o reconciliador usar `ReactDOMComponent` como a implementação da instância interna host.

Author: ```suggestion
* Renderizadores usam [injeção](/docs/codebase-overview.html#dynamic-injection) para passar a classe interna do host ao reconciliador. Por exemplo, o React DOM pede para o reconciliador usar `ReactDOMComponent` como a implementação da instância interna hospedeira.
```

> * A lógica para atualizar a lista de filhos é extraído em um mixin chamado `ReactMultiChild` que é usada pela implementação da classe de instância interna host tanto no React DOM quanto no React Native.

Author: ```suggestion
* A lógica para atualizar a lista de filhos é extraído em um mixin chamado `ReactMultiChild` que é usada pela implementação da classe de instância interna hospedeira tanto no React DOM quanto no React Native.
```

> * O Reconciliador também implementa suporte para `setState()` em elementos composite. Múltiplas atualizações dentro de event handlers são loteadas em uma so atualização.

Author: ```suggestion
* O Reconciliador também implementa suporte para `setState()` em elementos compostos. Múltiplas atualizações dentro de event handlers são loteadas em uma só atualização.
```

> * O reconciliador também lida com anexar e desanexar refs a componentes composite e nos host.

Author: ```suggestion
* O reconciliador também lida com anexar e desanexar refs a componentes compostos e nos hospedeiros.
```

* Métodos de ciclo de vida são chamados após o DOM estar pronto, como `componentDidMount()` e `componentDidUpdate()`, são coletados em "filas de callback" e são executadas em um só lote.

* O React coloca informações sobre a atualização atual no objeto interno chamado de "transação". Transações são úteis para observar a fila de métodos de ciclo de vida pendentes, o aninhamento do DOM atual para os warnings, e qualquer outra coisa que seja "global" a uma atualização específica. Transações também garantem que o React "limpe tudo" após atualizações. Por exemplo, a classe de transação provida pelo React DOM restaura a seleção de input após qualquer atualização.

> * React puts information about the current update into an internal object called "transaction". Transactions are useful for keeping track of the queue of pending lifecycle methods, the current DOM nesting for the warnings, and anything else that is "global" to a specific update. Transactions also ensure React "cleans everything up" after updates. For example, the transaction class provided by React DOM restores the input selection after any update.

Author: ```suggestion
```

### Mergulhando no código {#jumping-into-the-code}

* [`ReactMount`](https://github.com/facebook/react/blob/83381c1673d14cd16cf747e34c945291e5518a86/src/renderers/dom/client/ReactMount.js) é onde o código como `mountTree()` e `unmountTree()` deste tutorial vive.  Ele cuida da montagem e desmontagem de componentes de nível superior. [`ReactNativeMount`](https://github.com/facebook/react/blob/83381c1673d14cd16cf747e34c945291e5518a86/src/renderers/native/ReactNativeMount.js) é o análogo para React Native.

> * [`ReactDOMComponent`](https://github.com/facebook/react/blob/83381c1673d14cd16cf747e34c945291e5518a86/src/renderers/dom/shared/ReactDOMComponent.js) é o equivalente de `DOMComponent` neste tutorial. Ele implementa a componente de classe host para o renderizador React DOM.[`ReactNativeBaseComponent`](https://github.com/facebook/react/blob/83381c1673d14cd16cf747e34c945291e5518a86/src/renderers/native/ReactNativeBaseComponent.js) é o análogo para React Native.

Author: ```suggestion
* [`ReactDOMComponent`](https://github.com/facebook/react/blob/83381c1673d14cd16cf747e34c945291e5518a86/src/renderers/dom/shared/ReactDOMComponent.js) é o equivalente de `DOMComponent` neste tutorial. Ele implementa o componente de classe hospedeira para o renderizador React DOM.[`ReactNativeBaseComponent`](https://github.com/facebook/react/blob/83381c1673d14cd16cf747e34c945291e5518a86/src/renderers/native/ReactNativeBaseComponent.js) é o análogo para React Native.
```
